File: app/src/main/java/com/lineclient/home/homelineclient/activity/BaseActivity.java
package com.lineclient.home.homelineclient.activity;

import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;
import android.support.v4.app.FragmentActivity;
import android.widget.Toast;

import com.lineclient.home.homelineclient.application.MyApplication;
import com.lineclient.home.homelineclient.net.ViewInterface;
import com.lineclient.home.homelineclient.tools.FingerHelper;
import com.lineclient.home.homelineclient.view.FingerDialog;
import com.lineclient.home.homelineclient.view.LoadingDialog;

import java.lang.ref.WeakReference;

/**
 * Created by yangfengyuan on 2017/7/24.
 */

public class BaseActivity extends FragmentActivity implements ViewInterface {

    private boolean isPause = false;
    private boolean isUseLock = true;
    private MyApplication myApplication;
    private ViewHandler viewHandler;
    private LoadingDialog loadingDialog;

    private static class ViewHandler extends Handler {

        WeakReference<BaseActivity> weakReference;

        ViewHandler(BaseActivity object) {
            weakReference = new WeakReference<>(object);
        }

        @Override
        public void handleMessage(Message msg) {
            super.handleMessage(msg);

            BaseActivity object = weakReference.get();
            if (object != null) {
                if (!object.isPause) {
                    switch (msg.what) {
                        case 0:
                            Toast.makeText(object, (String) msg.obj, Toast.LENGTH_LONG).show();
                            break;
                        case 1:
                            Toast.makeText(object, (String) msg.obj, Toast.LENGTH_SHORT).show();
                            break;
                        case 3:
                            object.loadingDialog.show();
                            break;
                        case 4:
                            object.loadingDialog.dismiss();
                            break;
                    }
                }
            }

        }

    }

    @Override
    public void showLoading() {
        Message msg = new Message();
        msg.what = 3;
        viewHandler.sendEmptyMessage(3);
    }

    @Override
    public void closeLoading() {
        Message msg = new Message();
        msg.what = 3;
        viewHandler.sendEmptyMessage(4);
    }

    @Override
    public void showErrorMsg(String msg) {
        showLongToast(msg);
    }

    @Override
    public void showMsg(String msg) {
        showShortToast(msg);
    }

    public void setUseLock(boolean useLock) {
        isUseLock = useLock;
    }

    public void showLongToast(String text) {
        Message msg = new Message();
        msg.what = 0;
        msg.obj = text;
        viewHandler.sendMessage(msg);
    }

    public void showShortToast(String text) {
        Message msg = new Message();
        msg.what = 1;
        msg.obj = text;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        myApplication = (MyApplication) this.getApplication();
        viewHandler = new ViewHandler(this);
        initView();
    }

    private void initView() {
        loadingDialog = new LoadingDialog(this);
    }

    @Override
    protected void onPause() {
        super.onPause();
        isPause = true;
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (isUseLock) {
            if (isPause) {

                if (FingerHelper.checkUsed(this)) {

                    final FingerDialog fingerDialog = new FingerDialog(this);
                    fingerDialog.show();

                    FingerHelper.startCheckWithRadom(this.getApplicationContext(), new FingerHelper.FingerHelperInterfaceUnLock() {
                        @Override
                        public void success() {
                            fingerDialog.dismiss();
                        }

                        @Override
                        public void failed() {
                            fingerDialog.dismiss();
                        }

                        @Override
                        public void error(String msg) {
                            fingerDialog.dismiss();
                        }
                    });
                }

            }
        }
        isPause = false;
    }
}

File: app/src/main/java/com/lineclient/home/homelineclient/activity/BaseActivity.kt
package com.lineclient.home.homelineclient.activity

import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.support.v4.app.FragmentActivity
import android.widget.Toast
import com.lineclient.home.homelineclient.application.MyApplication
import com.lineclient.home.homelineclient.net.ViewInterface
import com.lineclient.home.homelineclient.tools.FingerHelper
import com.lineclient.home.homelineclient.tools.FingerHelper.FingerHelperInterfaceUnLock
import com.lineclient.home.homelineclient.view.FingerDialog
import com.lineclient.home.homelineclient.view.LoadingDialog
import java.lang.ref.WeakReference

/**
 * Created by yangfengyuan on 2017/7/24.
 */
open class BaseActivity : FragmentActivity(), ViewInterface {
    private var isPause = false
    private var isUseLock = true
    private var myApplication: MyApplication? = null
    private var viewHandler: ViewHandler? = null
    private var loadingDialog: LoadingDialog? = null

    private class ViewHandler internal constructor(objects: BaseActivity) : Handler() {
        var weakReference: WeakReference<BaseActivity> = WeakReference(objects)
        override fun handleMessage(msg: Message) {
            super.handleMessage(msg)
            val objects = weakReference.get()
            if (objects?.isPause==false) {
                when (msg.what) {
                    0 -> Toast.makeText(objects, msg.obj as String, Toast.LENGTH_LONG)
                            .show()
                    1 -> Toast.makeText(objects, msg.obj as String, Toast.LENGTH_SHORT)
                            .show()
                    3 -> objects.loadingDialog?.show()
                    4 -> objects.loadingDialog?.dismiss()
                }
            }
        }

    }

    override fun showLoading() {
        val msg = Message()
        msg.what = 3
        viewHandler?.sendEmptyMessage(3)
    }

    override fun closeLoading() {
        val msg = Message()
        msg.what = 3
        viewHandler?.sendEmptyMessage(4)
    }

    override fun showErrorMsg(msg: String?) {
        showLongToast(msg)
    }

    override fun showMsg(msg: String?) {
        showShortToast(msg)
    }

    fun setUseLock(useLock: Boolean) {
        isUseLock = useLock
    }

    fun showLongToast(text: String?) {
        val msg = Message()
        msg.what = 0
        msg.obj = text
        viewHandler?.sendMessage(msg)
    }

    fun showShortToast(text: String?) {
        val msg = Message()
        msg.what = 1
        msg.obj = text
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        myApplication = this.application as? MyApplication
        viewHandler = ViewHandler(this)
        initView()
    }

    private fun initView() {
        loadingDialog = LoadingDialog(this)
    }

    override fun onPause() {
        super.onPause()
        isPause = true
    }

    override fun onResume() {
        super.onResume()
        if (isUseLock) {
            if (isPause) {
                if (FingerHelper.checkUsed(this)) {
                    val fingerDialog = FingerDialog(this)
                    fingerDialog.show()
                    FingerHelper.startCheckWithRadom(
                            this.applicationContext, object : FingerHelperInterfaceUnLock {
                        override fun success() {
                            fingerDialog.dismiss()
                        }

                        override fun failed() {
                            fingerDialog.dismiss()
                        }

                        override fun error(msg: String) {
                            fingerDialog.dismiss()
                        }
                    })
                }
            }
        }
        isPause = false
    }
}